### DISCUSSION TITLE
HOW TO CLEAR THE VIEW？

### DISCUSSION THREAD

**Discussion by chaoyangxiao** [Kitware/trame#330] [Q&A] [answered]
Please see my code below, I want to use trame to be able to open a file to be displayed on the web side and to be able to clear the display on the web side, but the page got stuck after I clicked on the clear button, how exactly do I update this data change? I also want to have more operations on the data later, such as data excision, deformation, etc. I want to learn the mechanism of trame.

```python
server = get_server()
state = server.state

class MainWindow():
    def __init__(self):
        # -----------------------------------------------------------------------------
        # Trame initialization
        # -----------------------------------------------------------------------------

        self.ctrl = server.controller
        state.trame__title = "Web3D"

        # 设定初始值
        self.actor = vtkActor()
        cone_source = vtkConeSource()
        self.mapper = vtkPolyDataMapper()
        self.mapper.SetInputConnection(cone_source.GetOutputPort())
        self.actor.SetMapper(self.mapper)
        self.renderer = vtkRenderer()
        self.renderer.AddActor(self.actor)
        self.renderWindow = vtkRenderWindow()
        self.renderWindowInteractor = vtkRenderWindowInteractor()
        self.style = vtkInteractorStyleTrackballCamera()

        self.renderer.SetBackground(0.3, 0.3, 0.3)
        self.renderWindow.AddRenderer(self.renderer)
        self.renderWindowInteractor.SetRenderWindow(self.renderWindow)
        self.renderWindowInteractor.SetInteractorStyle(self.style)

        # 读取
        self.stl_reader = vtkSTLReader()

        # 布局
        with SinglePageLayout(server) as layout:
            layout.icon.click = self.ctrl.view_reset_camera
            layout.title.set_text("STL Application")

            with layout.toolbar:
                vtf_widgets.VDivider(vertical=True, classes="mx-2")
                # with vtf_widgets.VBtn(color="primary", icon=True):
                #     vtf_widgets.VIcon("mdi-arrow-split-vertical")

                with vtf_widgets.VBtn(color="primary", icon=True, click=self.openFile):
                    vtf_widgets.VIcon("mdi-folder-open")

                # remove
                with vtf_widgets.VBtn(color="primary", icon=True,click=self.clearScreen):
                    vtf_widgets.VIcon("mdi-close")

            with layout.content:
                with vtf_widgets.VContainer(
                        fluid=True,
                        classes="pa-0 fill-height",
                ):
                    self.view = vtk_widgets.VtkLocalView(self.renderWindow, ref="view")
                    self.ctrl.view_update = self.view.update
                    self.ctrl.view_reset_camera = self.view.reset_camera

        # 打开服务器
        server.start()


    def openFile(self):
        root = Tk()
        file_path = filedialog.askopenfilename()
        if file_path:
            dirname = os.path.dirname(file_path)
            basename = os.path.basename(file_path)

            print("目录路径:", dirname)
            print("文件名:", basename)

        root.destroy()
        file_p = os.path.join(dirname, basename)

        self.stl_reader.SetFileName(file_p)
        self.stl_reader.Update()

        self.mapper.SetInputData(self.stl_reader.GetOutput())
        self.actor.SetMapper(self.mapper)

        self.renderer.AddActor(self.actor)
        self.renderer.ResetCamera()

        self.renderWindow.Render()

        ## 更新窗口
        self.ctrl.view_update()
        self.ctrl.view_reset_camera()


    def clearScreen(self):

        # 清除屏幕
        self.renderer.RemoveAllViewProps()
        self.renderWindow.AddRenderer(self.renderer)
        self.renderWindowInteractor.SetRenderWindow(self.renderWindow)
        self.renderWindowInteractor.SetInteractorStyle(self.style)
        self.renderWindow.Render()

        self.ctrl.view_update()
        self.ctrl.view_reset_camera()

if __name__ == "__main__":
    mainwin = MainWindow()

```

**Comment by jourdain** [comment]
replace `VtkLocalView` by `VtkRemoteView`. Otherwise, if you must keep the local view, you need a dummy actor in your scene to properly keep the renderer and update its state.

Also no need to call `self.renderWindow.Render()` as that defeat the purpose of using the local view.

**Comment by jourdain** [comment, marked as answer]
Also that method should just be

```python
def clearScreen(self):
    self.renderer.RemoveAllViewProps()
    # if local view add a dummy actor/mapper/polydata in self.renderer
    self.ctrl.view_update()
```

**Comment by chaoyangxiao** [comment]
Thanks for your reply, I followed the method you said, replace VtkLocalView by VtkRemoteView.I have successfully cleared the actor.But since I am new to trame use and I have not done any web development work, I don't really understand the difference between VtkLocalView and VtkRemoteView. What is it? How should I use them? My main goal at the moment is to make a tool that can work with stl data on the web side, such as deforming, building, and a whole lot of other work on the model, I look forward to your reply, thank you for your help!
| |
xiaochaoyang0214
|
|
***@***.***
|
---- Replied Message ----
| From | Sebastien Jourdain ***@***.***> |
| Date | 9/13/2023 11:12 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [Kitware/trame] HOW TO CLEAR THE VIEW？ (Discussion #330) |

Also that method should just be

defclearScreen(self):
    self.renderer.RemoveAllViewProps()
    # if local view add a dummy actor/mapper/polydata in self.rendererself.ctrl.view_update()

**Comment by jourdain** [comment]
VtkLocalView partially support what VTK/C++ has to offer since it rely on vtk.js to render the data within the browser.
I would not suggest using the VtkLocalView to start with. While being a very powerful tool, you should only embrace it once you properly understand its implication.

If you want to read more on that subject, we have [that section](https://kitware.github.io/trame/docs/tutorial-vtk.html#Local-and-Remote-Rendering) in our tutorial.

HTH

**Comment by chaoyangxiao** [comment]
i got you.I'm using remoteview. i created a planewidget and it's stuck directly on the web side, it doesn't interact and the program crashes after several unresponsive clicks.
def addplaneWidget(self):

# 新建一个平面
planeWidget = vtk.vtkPlaneWidget()

    planeWidget.SetInteractor(self.renderWindowInteractor)
    planeWidget.SetPlaceFactor(1.0)
    planeWidget.SetResolution(50)
    planeWidget.SetRepresentationToSurface()
    planeWidget.On()

self.ctrl.view_update()
self.ctrl.view_reset_camera()




| |
xiaochaoyang0214
|
|
***@***.***
|
---- Replied Message ----
| From | Sebastien Jourdain ***@***.***> |
| Date | 9/13/2023 11:34 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [Kitware/trame] HOW TO CLEAR THE VIEW？ (Discussion #330) |

VtkLocalView partially support what VTK/C++ since it rely on vtk.js to render the data within the browser.
Unless you understand what you are doing, I would not suggest using the VtkLocalView.

If you want to read more on that subject, we have that section in our tutorial.

**Comment by jourdain** [comment]
https://examples.vtk.org/site/Python/Widgets/ImplicitPlaneWidget2/

**Comment by chaoyangxiao** [comment]
Hello, I have been learning trame for a few days now and I understand that using localview involves synchronization of client and server side data, I think this is a bit difficult and troublesome for me. So I decided to use remoteview, this way I can develop QT applications as easy as possible, but during my use I found that this way rendering and processing data will be very slow, for example if I add a planewidget, the web page will be stuck and the program will crash and can't complete the interaction. Is there any solution for this? Or is there any way to make remoteview faster to render and process data?
     Looking forward to your reply.


| |
xiaochaoyang0214
|
|
***@***.***
|
---- Replied Message ----
| From | Sebastien Jourdain ***@***.***> |
| Date | 9/13/2023 21:38 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [Kitware/trame] HOW TO CLEAR THE VIEW？ (Discussion #330) |

https://examples.vtk.org/site/Python/Widgets/ImplicitPlaneWidget2/

**Comment by jourdain** [comment]
I don't get the relationship with Qt and trame. Also I have no clue what you are doing for making your code slow and crash. The only thing I can say is that when we use 3D widgets with remote view, we don't experience what you are describing.

**Comment by chaoyangxiao** [comment]
Hello, I found the reason why my 3dwidget can't interact in remoteview, I have to initialize a 3dwidget in a global variable and then trigger it by a button click event, this way i can use the 3dwidget just fine,for example
planeWidget = vtk.vtkPlaneWidget()
planeWidget.SetInputConnection(stl_reader.GetOutputPort())
planeWidget.SetRepresentationToSurface()
planeWidget.SetInteractor(renderWindowInteractor)
planeWidget.SetPlaceFactor(1.0)
planeWidget.SetResolution(50)
def addplanewidget():
    planeWidget.PlaceWidget()
    planeWidget.On()
    ctrl.view_update()
    ctrl.view_reset_camera()
When I add the 3dwidget as per the following function, the web interface crashes straight away.
def addplanewidget():
    planeWidget = vtk.vtkPlaneWidget()
    planeWidget.SetInputConnection(stl_reader.GetOutputPort())
    planeWidget.SetRepresentationToSurface()
    planeWidget.SetInteractor(renderWindowInteractor)
    planeWidget.SetPlaceFactor(1.0)
    planeWidget.SetResolution(50)
    planeWidget.PlaceWidget()
    planeWidget.On()

    ctrl.view_update()
    ctrl.view_reset_camera()
Thank you for your patience in solving my problem, and again, I ask for your advice as to why the problem occurs as I have described.

| |
xiaochaoyang0214
|
|
***@***.***
|
---- Replied Message ----
| From | Sebastien Jourdain ***@***.***> |
| Date | 9/16/2023 22:50 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [Kitware/trame] HOW TO CLEAR THE VIEW？ (Discussion #330) |

I don't get the relationship with Qt and trame. Also I have no clue what you are doing for making your code slow and crash. The only thing I can say is that when we use 3D widgets with remote view, we don't experience what you are describing.

**Comment by jourdain** [comment]
Without a proper reproducible example, we can't help you.
